copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+			if hdr.Typeflag == tar.TypeDir {
+				if request.PutOptions.ChmodDirs != nil {
+					hdr.Mode = int64(*request.PutOptions.ChmodDirs)
+				}
+			} else {
+				if request.PutOptions.ChmodFiles != nil {
+					hdr.Mode = int64(*request.PutOptions.ChmodFiles)
+				}
+			}
+			// create the new item
+			devMajor := uint32(hdr.Devmajor)
+			devMinor := uint32(hdr.Devminor)
+			written := hdr.Size
+			switch hdr.Typeflag {
+			// no type flag for sockets
+			case tar.TypeReg, tar.TypeRegA:
+				written, err = createFile(path, tr, mode)
+			case tar.TypeLink:
+				var linkTarget string
+				if filepath.IsAbs(hdr.Linkname) {
+					linkTarget, err = resolvePath(targetDirectory, filepath.Join(request.Root, hdr.Linkname), nil)
+					if err != nil {
+						return errors.Errorf("error resolving hardlink target path %q under root %q", hdr.Linkname, request.Root)
+					}
+				} else {
+					linkTarget, err = resolvePath(targetDirectory, filepath.Join(targetDirectory, filepath.Dir(hdr.Name), hdr.Linkname), nil)
+					if err != nil {
+						return errors.Errorf("error resolving hardlink target path %q under root %q in directory %q", hdr.Linkname, request.Root, filepath.Dir(hdr.Name))
+					}
+				}
+				err = os.Link(linkTarget, path)
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = os.Link(linkTarget, path)
+					}
+				}
+			case tar.TypeSymlink:
+				err = os.Symlink(hdr.Linkname, path)
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = os.Symlink(hdr.Linkname, path)
+					}
+				}
+			case tar.TypeChar:
+				err = mknod(path, chrMode(mode), int(mkdev(devMajor, devMinor)))
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = mknod(path, chrMode(mode), int(mkdev(devMajor, devMinor)))
+					}
+				}
+			case tar.TypeBlock:
+				err = mknod(path, blkMode(mode), int(mkdev(devMajor, devMinor)))
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = mknod(path, blkMode(mode), int(mkdev(devMajor, devMinor)))
+					}
+				}
+			case tar.TypeDir:
+				err = os.Mkdir(path, mode)
+				if err != nil && os.IsExist(err) {
+					err = nil
+				}
+				// make a note of the directory's times.  we
+				// might create items under it, which will
+				// cause the mtime to change after we correct
+				// it, so we'll need to correct it again later
+				directoriesAndTimes = append(directoriesAndTimes, directoryAndTimes{
+					directory: path,
+					atime:     hdr.AccessTime,
+					mtime:     hdr.ModTime,
+				})
+			case tar.TypeFifo:
+				fifoMode := uint32(hdr.Mode)
+				err = mkfifo(path, fifoMode)
+				if err != nil && os.IsExist(err) {
+					if err = os.Remove(path); err == nil {
+						err = mkfifo(path, fifoMode)
+					}
+				}
+			}
+			// check for errors
+			if err != nil {
+				return errors.Wrapf(err, "error creating %q", path)
+			}
+			if written != hdr.Size {
```

Comment:
(It might be simpler to check for this only in the `TypeReg` case.)

Reply:
Sure, changed it.